docs/proposals/config_data.md

```diff
@@ -0,0 +1,327 @@
+# Generic Configuration Object
+
+## Abstract
+
+This proposal proposes a generic configuration object named ``ConfigData`` stored inside ``etcd`` that
+stores data used for the configuration of ``Kubernetes`` components.
+
+The main focus points of this proposal are:
+
+* Solving dynamic configuration problem of components.
+* Encapsulate configuration information to decrease component configuration complexity.
+* Create a flexible configuration model for ``Kubernetes`` components.
+* Increase overall system quality.
+
+## Motivation
+
+Currently command-line flags and environment variables are used for component configurations in ``Kubernetes``
+which is complicated and error-prone. This approach is not suitable for a distributed computing environment.
+Specifically the problems with the current approach are:
+
+* Complicated deployment orchestration
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+* Hard to change in different versions that may support different configuration options.
+
+## Use Cases
+
+1. Be able to load API resources as environment variables into a component.
+2. Be able to load API resources as mounted volumes into a component.
+3. Be able to update mounted volumes after a change to the API resource.
+4. Be able to configure components using an API resource instead of command-line flags.
+
+## Solution
+
+A new ``ConfigData`` API resource is proposed to address these problems.
+
+The ``ConfigData`` architecture will be very similar to that of ``Secrets``.
+
+``ConfigData`` can either be mounted as a volume or can be injected as environment
+variables to a container through the use of the downward API.
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container (downward API).
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the API server every time the object is modified.
+After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs
+only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsibility to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object).
```

Comment:
@krancour We had a discussion about this in the community hangout yesterday.  The problem with automatically restarting pods is that not all pods will have the same wants.  We also want to avoid coupling to kubernetes features.  An init-container in a pod has been proposed as a way to straddle the right line between making updates available (volume plugin will be eventually consistent) and being agnostic of how applications want to handle updates to configs (init-container can monitor configs and restart affected containers).

@bgrant0607 @brendandburns @smarterclayton am I articulating that correctly?